### ISSUE TITLE
borg under windows 10's linux subsystem

### ISSUE THREAD

**Issue by ThomasWaldmann** [borgbackup/borg#1961]
try it, test it, document outcome.

**Comment by robaato** [comment]
First impression: 
borg 1.0.8 from bundled binary borg-linux64 -> starts fine

mkdir aaa
./borg-linux64 init aaa
./borg-linux64 create --info --progress aaa::bbb /usr/
./borg-linux64 list aaa::bbb
./borg-linux64 check aaa
./borg-linux64 extract aaa::bbb
diff -rN --brief /usr usr  
// no result except for few broken links, which are already broken on source

all above run fine. Looks promising. Will try to backup and restore something from Windows partitions, above is all inside linux subsystem.

**Comment by ThomasWaldmann** [comment]
@robaato thanks for the update. btw, you don't need to create the repo dir first, borg init does that, too.

**Comment by ThomasWaldmann** [comment]
My first (install via pip from pypi package) and second impression (install from git repo) is also that it works.

Installation is like on (native) Ubuntu (Trusty, 14.04). Only thing I noticed not working is the FUSE support (borg mount) as it requires the fuse kernel module, which is not there.

Now running the tests...

**Comment by ThomasWaldmann** [comment]
23 test fails:
- create_test_files: os.mknod -> Function not implemented (ENOSYS)
- get_acl (from linux platform module) returns dict without b'acl_access' key
- fuse tests fail (no fuse kernel support)

**Comment by ThomasWaldmann** [comment]
See PR #1962.

**Comment by BluCodeGH** [comment]
I have run borg 1.0.9 under the windows 10 subsystem as an administrator and while using sudo. I am backing up files from my (windows) home folder to another drive. The only errors I am seeing are [Errno 13] Permission Denied, and [Errno 5] Input/output error. Error 13 occurs on some files and folders, such as AppData/Local/ElevatedDiagnostics or Appdata/Local/Google/Chrome/User Data/Default/Current Tabs, and error 5 occures on files called LOCK, in various folders.

**Comment by enkore** [comment]
Locking errors on Windows are to be expected unless VSS is used.

**Comment by BluCodeGH** [comment]
Is there a way to get borg on lxss working with VSS?

**Comment by jdchristensen** [comment]
And any idea why some files (like ElevatedDiagnostics and Current Tabs) get Errno 13, while the LOCK files get Errno 5?  The LOCK files are probably not important, but it would be good if the other files could be backed up.

**Comment by PythonNut** [comment]
> Is there a way to get borg on lxss working with VSS?

I just gave this a shot, and as far as I can tell, the answer is no, since WSL doesn't support NTFS Junctions, which is how VSS snapshots are "mounted". This is indeed unfortunate.

**Comment by Clete2** [comment]
I am running under Windows 10 subsystem backing up 1.5TB of data without any errors. `borg check` on the remote host is also reporting that the integrity is good. Is there anything else I should check to ensure that the backup is complete and accurate?

FYI -- I am backing up video game downloads from the following game downloading services: Steam, Blizzard (formerly Battle.net), Origin, and some other misc games without downloading services.

**Comment by ThomasWaldmann** [comment]
borg check (in full mode, not just in --repository-only mode) being ok is quite a good sign already.

are more practical test is to actually extract all data again and compare to the original data.
or extract --dry-run if you do not have the storage space.

**Comment by mnpenner** [comment]
Assuming Borg works under WSL/Win10Bash, how would we get it to start automatically on Windows startup?

**Comment by enkore** [comment]
Perhaps this helps: https://superuser.com/a/1112022/107877

**Comment by ThomasWaldmann** [comment]
Well, maybe a matter of taste, but I find anything major starting automatically on startup (or shutdown) rather annoying.

**Comment by mnpenner** [comment]
@ThomasWaldmann  I do too, but we're talking about backup here. That has to run continuously, or we're vulnerable, no?

Thanks @enkore I'll take a look.

**Comment by sdellenb** [comment]
😃 Backup works surprisingly well with WSL, though there's not much data yet in the user's Windows folders.
I've configured the Windows Task Scheduler to run bash (`C:\Windows\System32\bash.exe`) with a wrapper script (arguments `-c "~/bin/borgmatic_wrapper.sh"`) daily at 18:00, with the option "Run task as soon as posible after a schedule start is missed".
I'll see how the setup holds up in the future.

Note: To get borgbackup 1.0.10, I had to install it from an [alternative PPA](https://launchpad.net/~costamagnagianfranco/+archive/ubuntu/borgbackup). I also installed `borgmatic` for easier configuration and my wrapper script collects the borgmatic output and mails it with `msmtp` (on Linux machines, I let `cron` handle this).

**Comment by rbn920** [comment]
I used the linux64 binary today and seems to work just fine.

**Comment by theChaosCoder** [comment]
I just wanted to report, that you can now mount network shares under WSL with the new Windows 10 Fall Creators Update. No more cygwin needed.

`$ sudo mount -t drvfs '\\server\share' /mnt/share`

See also https://blogs.msdn.microsoft.com/wsl/2017/04/18/file-system-improvements-to-the-windows-subsystem-for-linux/

**Comment by theChaosCoder** [comment]
I need to correct myself:

it worked with my small repo. But I get often this error:

```
Local Exception
Traceback (most recent call last):
  File "borg/archiver.py", line 4024, in main
  File "borg/archiver.py", line 3952, in run
  File "borg/archiver.py", line 148, in wrapper
  File "borg/archiver.py", line 528, in do_create
  File "borg/archiver.py", line 494, in create_inner
  File "borg/archive.py", line 446, in save
  File "borg/archive.py", line 250, in flush
  File "borg/archive.py", line 266, in write_chunk
  File "borg/cache.py", line 872, in add_chunk
  File "borg/crypto/key.py", line 362, in encrypt
  File "borg/crypto/nonces.py", line 85, in ensure_reservation
  File "borg/crypto/nonces.py", line 48, in commit_repo_nonce_reservation
  File "borg/repository.py", line 309, in commit_nonce_reservation
  File "borg/platform/base.py", line 170, in __exit__
  File "borg/platform/base.py", line 138, in close
  File "borg/platform/base.py", line 58, in sync_dir
OSError: [Errno 5] Input/output error

Platform: Linux Chaos-PC 4.4.0-43-Microsoft #1-Microsoft Wed Dec 31 14:42:53 PST 2014 x86_64 x86_64
Linux: debian stretch/sid
Borg: 1.1.0  Python: CPython 3.5.4
PID: 47  CWD: /mnt/borgBackup
```

Even borg init triggers OSError: [Errno 5] Input/output error.

I remember having similar problems with borg 1.0.10 in cygwin, but 1.0.9 was ok. I will investigate further.

**Comment by ThomasWaldmann** [comment]
Often or always? Maybe syncing a dir is not supported on windows?

**Comment by theChaosCoder** [comment]
always

**Comment by theChaosCoder** [comment]
Is there an easy way to test sync?

**Comment by ThomasWaldmann** [comment]
```
import errno
import os
import sys

def sync_dir(path):
    fd = os.open(path, os.O_RDONLY)
    try:
        os.fsync(fd)
    except OSError as os_error:
        # Some network filesystems don't support this and fail with EINVAL.
        # Other error codes (e.g. EIO) shouldn't be silenced.
        if os_error.errno != errno.EINVAL:
            raise
    finally:
        os.close(fd)

if __name__ == '__main__':
    path = sys.argv[1]
    print("Syncing dir %s" % path)
    sync_dir(path)
```

Usage:
```
python3 syncdir.py somedirectory
```

**Comment by theChaosCoder** [comment]
Syncing dir /mnt/borgBackup/
Traceback (most recent call last):
  File "stest.py", line 20, in <module>
    sync_dir(path)
  File "stest.py", line 8, in sync_dir
    os.fsync(fd)
OSError: [Errno 5] Input/output error

**Comment by ThomasWaldmann** [comment]
so, the "linux filesystem" syncs dirs ok (in WSL), but the mounted drvfs always fails with the dir sync?

**Comment by theChaosCoder** [comment]
exactly.

**Comment by theChaosCoder** [comment]
Also a mounted usb stick works too

sudo mount -t drvfs D: /mnt/Ddrive/ <-- OK
sudo mount -t drvfs F: /mnt/usbstick/ <-- OK
sudo mount -t drvfs '\\synology\Backups'  /mnt/borgBackup/ <-- Fail

**Comment by ThomasWaldmann** [comment]
\synology\Backups is what? smbfs via drvfs?

Guess that's a bug in one of these filesystems.
They should either support dir sync or say "invalid", but not "I/O error".

**Comment by theChaosCoder** [comment]
yep it is smbfs via drvfs?

**Comment by ThomasWaldmann** [comment]
Can you report this to microsoft?
We can't suppress EIO, it could be a serious error.

**Comment by enkore** [comment]
FlushFileBuffers does not support directory handles, therefore I'd guess that the kernel doesn't contain that functionality - this looks like a bad mapping of error codes (ERROR_INVALID_HANDLE => EIO instead of EINVAL or EBADF).

**Comment by bj0** [comment]
just tried with a freshly created repo, and get: 

Local Exception
Traceback (most recent call last):
  File "borg/archiver.py", line 4024, in main
  File "borg/archiver.py", line 3952, in run
  File "borg/archiver.py", line 148, in wrapper
  File "borg/archiver.py", line 528, in do_create
  File "borg/archiver.py", line 492, in create_inner
  File "borg/archiver.py", line 593, in _process
  File "borg/archiver.py", line 593, in _process
  File "borg/archiver.py", line 593, in _process
  File "borg/archiver.py", line 593, in _process
  File "borg/archiver.py", line 593, in _process
  File "borg/archiver.py", line 593, in _process
  File "borg/archiver.py", line 593, in _process
  File "borg/archiver.py", line 571, in _process
  File "borg/archive.py", line 991, in process_file
  File "borg/archive.py", line 919, in chunk_file
  File "borg/archive.py", line 911, in chunk_processor
  File "borg/cache.py", line 874, in add_chunk
  File "borg/repository.py", line 1054, in put
  File "borg/repository.py", line 1424, in write_put
  File "src/borg/platform/linux.pyx", line 248, in borg.platform.linux.SyncFile.write
  File "src/borg/platform/linux.pyx", line 222, in borg.platform.linux._sync_file_range
OSError: [Errno 38] Function not implemented

Platform: Linux  4.4.0-43-Microsoft #1-Microsoft Wed Dec 31 14:42:53 PST 2014 x86_64 x86_64
Linux: debian stretch/sid
Borg: 1.1.0  Python: CPython 3.5.4

**Comment by ThomasWaldmann** [comment]
@bj0 looks like they didn't implement this yet:
```
    int sync_file_range(int fd, int64_t offset, int64_t nbytes, unsigned int flags)
```

**Comment by henfri** [comment]
Hello,

I am trying this under ubuntu (in windows 10). when creating a backup, I get:
Local Exception.
```
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/borg/archiver.py", line 2168, in main
    exit_code = archiver.run(args)
  File "/usr/lib/python3/dist-packages/borg/archiver.py", line 2104, in run
    return set_ec(func(args))
  File "/usr/lib/python3/dist-packages/borg/archiver.py", line 101, in wrapper
    kwargs['manifest'], kwargs['key'] = Manifest.load(repository, compatibility)
  File "/usr/lib/python3/dist-packages/borg/helpers.py", line 253, in load
    data = key.decrypt(None, cdata)
  File "/usr/lib/python3/dist-packages/borg/key.py", line 281, in decrypt
    data = self.compressor.decompress(self.dec_cipher.decrypt(data[41:]))
  File "borg/compress.pyx", line 210, in borg.compress.Compressor.decompress (borg/compress.c:4825)
ValueError: ('No decompressor for this data found: %r.', b'\x03\x00')
```
I have installed python3-lz4
Borg is installed via apt.

What is missing?

Greetings,
Hendrik

**Comment by henfri** [comment]
With this release it works:
https://github.com/borgbackup/borg/releases

**Comment by ThomasWaldmann** [comment]
0300 is zstd compressed data, you need at least borg 1.1.4 for that.

**Comment by henfri** [comment]
Thanks. In fact I was aware of the fact that zstd needs 1.1.4, but I was not using zstd for this job.
But in this archive on the Server I have jobs that did use zstd.

**Comment by ThomasWaldmann** [comment]
The problem is that the Manifest (contains, among other stuff, the list of all archives) is compressed with zstd. BTW, this is not specific to WSL, so it is misplaced in this ticket.

**Comment by henfri** [comment]
I understand that -now. 
I was not aware yesterday though.

Greetings,
Hendrik

**Comment by henfri** [comment]
Hello again,

unfortunately, I get many 'connection closed by remote host' messages. I have not yet been able to complete a backup. The connection is usually lost after around one hour (but varies).
After that, the lock is still on the Server and I need to break it manually.

Are there any options to make borg do a retry?


Greetings,
Hendrik

**Comment by ThomasWaldmann** [comment]
@henfri no. but you rather should fix the root cause anyway. maybe ssh or router disconnecting, or other connection issue?

About the server lock: did you check whether the server process ("borg serve") has terminated?

**Comment by henfri** [comment]
Hello Thomas,

thanks for your reply.
The Server has been quite stable. I did a full extract test recently from my linux client. Of course I cannot exclude that this has changed since..

Regarding the server process:
I am not sure. I thought that the server process is started via ssh by the client?
Below you find a ps of when I started the backup and now (the backup has ended with connection closed again).

At start of backup:
```
  PID COMMAND                          STARTED     ELAPSED     TIME COMMAND
henfri@openmediavault:~$ ps -eo pid,comm,lstart,etime,time,args | grep borg
22362 sshd            Sat Mar 10 15:56:53 2018       22:04 00:00:00 sshd: borg [priv]
22365 sshd            Sat Mar 10 15:56:53 2018       22:04 00:00:35 sshd: borg@notty
22368 borg            Sat Mar 10 15:56:54 2018       22:03 00:00:01 borg serve --restrict-to-path /srv/_dev_disk_by-id_ata-ST5000DM000-1FK178_W4J1K7R1
22369 borg            Sat Mar 10 15:56:55 2018       22:02 00:00:19 borg serve --restrict-to-path /srv/_dev_disk_by-id_ata-ST5000DM000-1FK178_W4J1K7R1
11089 sshd            Sat Mar 10 13:17:47 2018    03:01:10 00:00:00 sshd: borg [priv]
11091 sshd            Sat Mar 10 13:17:47 2018    03:01:10 00:13:43 sshd: borg@notty
11092 borg            Sat Mar 10 13:17:47 2018    03:01:10 00:00:01 borg serve --restrict-to-path /srv/_dev_disk_by-id_ata-ST5000DM000-1FK178_W4J1K7R1
11093 borg            Sat Mar 10 13:17:49 2018    03:01:08 00:04:53 borg serve --restrict-to-path /srv/_dev_disk_by-id_ata-ST5000DM000-1FK178_W4J1K7R1

```

Now:
```
  PID COMMAND                          STARTED     ELAPSED     TIME COMMAND
henfri@openmediavault:~$ ps -eo pid,comm,lstart,etime,time,args | grep borg
22362 sshd            Sat Mar 10 15:56:53 2018    01:39:23 00:00:00 sshd: borg [priv]
22365 sshd            Sat Mar 10 15:56:53 2018    01:39:23 00:00:35 sshd: borg@notty
22368 borg            Sat Mar 10 15:56:54 2018    01:39:22 00:00:01 borg serve --restrict-to-path /srv/_dev_disk_by-id_ata-ST5000DM000-1FK178_W4J1K7R1
22369 borg            Sat Mar 10 15:56:55 2018    01:39:21 00:00:19 borg serve --restrict-to-path /srv/_dev_disk_by-id_ata-ST5000DM000-1FK178_W4J1K7R1
24109 sshd            Sat Mar 10 16:19:22 2018    01:16:54 00:00:00 sshd: borg [priv]
24111 sshd            Sat Mar 10 16:19:22 2018    01:16:54 00:07:43 sshd: borg@notty
24112 borg            Sat Mar 10 16:19:22 2018    01:16:54 00:00:01 borg serve --restrict-to-path /srv/_dev_disk_by-id_ata-ST5000DM000-1FK178_W4J1K7R1
24113 borg            Sat Mar 10 16:19:24 2018    01:16:52 00:02:41 borg serve --restrict-to-path /srv/_dev_disk_by-id_ata-ST5000DM000-1FK178_W4J1K7R1

```

The PID 
22362 sshd
22365 sshd
22368 borg
22369 borg

have kept running, while 
11089 sshd 
11091 sshd 
11092 borg 
11093 borg 

have stopped and are seemingly replaced by
24109 sshd
24111 sshd
24112 borg
24113 borg

**Comment by henfri** [comment]
Hello Thomas,

thanks. Before I could do this, the backup did run through. I have changed nothing (TM). The only difference was, that I did run a ping with the parameter -i 0.2 to the server in parallel. I would not expect this to have an effect though.

Now, I have tried to do what you wrote above. Following findings:
1) before the backup, borg does *not* run on the server
2) during the backup I see these processes:
```
31617 sshd            Sun Mar 11 09:08:36 2018       00:10 00:00:00 sshd: borg [priv]
31622 sshd            Sun Mar 11 09:08:36 2018       00:10 00:00:00 sshd: borg@notty
31623 borg            Sun Mar 11 09:08:37 2018       00:09 00:00:01 borg serve --restrict-to-path /srv/Shortended
31626 borg            Sun Mar 11 09:08:38 2018       00:08 00:00:06 borg serve --restrict-to-path /srv/Shortened
```
3) The backup finishes successfully. Certainly the fact that there is little to backup helps here.... no prove that anything is fixed now.

Off topic: I have a question on this:

> But I guess after some timeout borg serve would notice that the connection has broken down and would terminate itself and also remove the repo lock.

Borg is run via ssh. If the connection breaks down, wouldn't borg be stopped in a way that it cannot remove the lock anymore?

Totally OT:
Do you plan to get rid of the lock in future and work with some mechanism (e.g. DB) that allows parallel access to a repository? Of course, one can setup one repository for each client, but then one cannot profit of the deduplication across clients.

Greetings,
Hendrik

**Comment by ThomasWaldmann** [comment]
We try to use context managers everywhere, so even if an exception happens, stuff (like locks) gets cleaned up / teared down in a orderly manner.

Long term it would be nice to get rid of the locking, but don't hold your breath, there are no specific plans for the near future.

**Comment by TheJP** [comment]
I got it to work by installing from git (tag 1.1.5) and commenting out this line https://github.com/borgbackup/borg/blob/master/src/borg/platform/__init__.py#L24 (`from .linux import SyncFile`)

I had the same problem as @bj0. See the referenced issue for some more details.

**Comment by ThomasWaldmann** [comment]
@TheJP nice trick to get it working. it syncs a bit less then (not on write, only on close/sync), but if that makes it work...

**Comment by TheJP** [comment]
That's how I understood it yes. But I rather have it working a bit inefficient than not at all. It did run successfully on a couple hundred GB and some 100'000 files by now.

Great project btw :)

**Comment by ThomasWaldmann** [comment]
Not sure it is less efficient. It is just not as quickly making sure write-data is actually on the media.
As more is cached and sync is happening later, writes might be a bit more bursty.

**Comment by ThomasWaldmann** [comment]
@TheJP do you know how one can detect WSL (opposed to normal Linux)? Hmm, we could just try this call...

**Comment by TheJP** [comment]
I didn't know a way, but it looks like you could check for "Microsoft" in `/proc/version` (https://github.com/Microsoft/WSL/issues/423).
A bit hacky, but not problematic in this case I think:
* A false positive would lead to the usage of the old `SyncFile` functionality on an actual linux system. No problem and not likely that another Linux distribution would have "Microsoft" in its `/proc/version`.
* A false negative would produce the same result as the current Borg version. (An Exception in my case.)

**Comment by intdel** [comment]
@ThomasWaldmann You can detect it through the kernel name:
Linux Computername 4.4.0-17134-Microsoft #1-Microsoft Tue Apr 10 18:04:00 PST 2018 x86_64 x86_64 x86_64 GNU/Linux

**Comment by henfri** [comment]
@TheJP 

> I got it to work by installing from git (tag 1.1.5) and commenting out this line https://github.com/borgbackup/borg/blob/master/src/borg/platform/__init__.py#L24 (from .linux import SyncFile)

How did you install from git?
when running pip install -e . 
I get this error:
```
    writing entry points to src/borgbackup.egg-info/entry_points.txt
    /usr/lib/python3.5/distutils/dist.py:261: UserWarning: Unknown distribution option: 'python_requires'
      warnings.warn(msg)
    warning: manifest_maker: standard file '-c' not found

    reading manifest template 'MANIFEST.in'
    writing manifest file 'src/borgbackup.egg-info/SOURCES.txt'
    running build_ext
    building 'borg.platform.linux' extension
    x86_64-linux-gnu-gcc -pthread -DNDEBUG -g -fwrapv -O2 -Wall -Wstrict-prototypes -g -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -fPIC -I/usr/include/python3.5m -c src/borg/platform/linux.c -o build/temp.linux-x86_64-3.5/src/borg/platform/linux.o
    src/borg/platform/linux.c:585:21: fatal error: sys/acl.h: No such file or directory
    compilation terminated.
    error: command 'x86_64-linux-gnu-gcc' failed with exit status 1

    ----------------------------------------
```

I am sure I can fix this, but then I will get the next error, so I wonder, whether I am missing something bigger.

Greetings,
Hendrik

**Comment by henfri** [comment]
sudo -H apt-get install libacl1-dev libssl-dev

fixed it for me.

**Comment by laurensdijkstra** [comment]
I also get
```
  File "src/borg/platform/linux.pyx", line 255, in borg.platform.linux.SyncFile.write
  File "src/borg/platform/linux.pyx", line 229, in borg.platform.linux._sync_file_range
OSError: [Errno 38] Function not implemented
```
Reading the comments here I understand it is due to a function not being implemented on the WSL side. The workaround could be commenting out a line from the src. How does one go about achieving that goal?

And does disabling this feature mean that my VM image will be copied over in its entirety every time a small change is made to it?

**Comment by laurensdijkstra** [comment]
It has already been reported to MS 2 years ago in https://github.com/Microsoft/WSL/issues/645. @TheJP referenced this issue before. Doesn't seem like a lot is happening to fix it.

**Comment by ks07** [comment]
So I've just been experimenting with borg in WSL and ran into the `sync_file_range` bug - but thankfully it appears this only affects repos hosted on Windows, and doesn't trigger if the repo is hosted remotely on Linux. It also seems to only trigger if the backup source is sufficiently large (presumably the sync is only triggered once enough data has been written?) . Am I right in thinking this? Or is this bug likely to be triggered even if the remote repo is Linux based?

**Comment by stephendt** [comment]
Just chiming in to say that I appear to be affected by this bug as well. Unfortunately I'm using Storage Spaces formatted as NTFS as my repo. I am on 1.1.5. Will check back to see if a fix arrives.

**Comment by ThomasWaldmann** [comment]
the problem with sync_file_range is worked around by #4317.

i'ld appreciate it if somebody who uses WSL could practically test it (i currently do not have a WSL installation, so it is completely untested by me for that platform).

basically this is automating what some of you did by commenting out the SyncFile import.

**Comment by ThomasWaldmann** [comment]
the WSL broken sync_file_range workaround is now in master and 1.1-maint (and will be in 1.1.9).

**Comment by ThomasWaldmann** [comment]
would be good if some people add their experiences with borg 1.1.9 under WSL below.

**Comment by brummelte** [comment]
https://launchpad.net/~costamagnagianfranco/+archive/ubuntu/borgbacku is still at 1.1.8. I will test it when the new version is available.

**Comment by LocutusOfBorg** [comment]
done.

**Comment by brummelte** [comment]
Thanks, I've tested 4 backup repositories now. All of them work. Previously only the small ones sometimes worked with WSL.

Small before (cygwin):
```
borgbackup version 1.1.7
------------------------------------------------------------------------------
Archive name: Jan-Jan-2019-02-20_07-47-45
Archive fingerprint: 5a165138db025188fcf7b4b46394d4d6cecd1404fbcac6d6ba80e2a4d7411c32
Time (start): Wed, 2019-02-20 07:47:56
Time (end):   Wed, 2019-02-20 07:48:19
Duration: 22.88 seconds
Number of files: 9824
Utilization of max. archive size: 0%
------------------------------------------------------------------------------
                       Original size      Compressed size    Deduplicated size
This archive:                4.27 GB              3.90 GB            143.29 kB
All archives:                4.69 TB              4.26 TB              6.04 GB

                       Unique chunks         Total chunks
Chunk index:                   47251             12761242
------------------------------------------------------------------------------
```

Small after (WSL):
```
borgbackup version 1.1.9
Creating archive at …
------------------------------------------------------------------------------
Archive name: Jan-jan-2019-02-20_09-56-46
Archive fingerprint: fbb89aaeb6e8480504152bae92538c25ada9631beb5ffb9ab12191ac6178dfde
Time (start): Wed, 2019-02-20 09:56:46
Time (end):   Wed, 2019-02-20 09:57:05
Duration: 18.69 seconds
Number of files: 9899
Utilization of max. archive size: 0%
------------------------------------------------------------------------------
                       Original size      Compressed size    Deduplicated size
This archive:                4.27 GB              3.90 GB            193.30 kB
All archives:                4.70 TB              4.26 TB              6.04 GB

                       Unique chunks         Total chunks
Chunk index:                   47299             12784076
------------------------------------------------------------------------------
```

Big before (cygwin):
```
borgbackup version 1.1.7
------------------------------------------------------------------------------
Archive name: Jan-Jan-2019-02-04_08-55-23
Archive fingerprint: 962bcaeeb95d2cd612a06af6f7c3b8bd1fbdf685d9bf5981f623eadf77c868d6
Time (start): Mon, 2019-02-04 08:55:26
Time (end):   Mon, 2019-02-04 09:45:00
Duration: 49 minutes 33.90 seconds
Number of files: 1312181
Utilization of max. archive size: 1%
------------------------------------------------------------------------------
                       Original size      Compressed size    Deduplicated size
This archive:               70.28 GB             38.48 GB              1.83 GB
All archives:                9.83 TB              4.55 TB            109.95 GB

                       Unique chunks         Total chunks
Chunk index:                 1392093            151273982
------------------------------------------------------------------------------
```

Big after (WSL):
```
borgbackup version 1.1.9
Creating archive at …
------------------------------------------------------------------------------
Archive name: Jan-jan-2019-02-20_15-34-06
Archive fingerprint: 06ff40c8d51554618b7661543f005eb6e1750e772a7996d75319213f370b29b2
Time (start): Wed, 2019-02-20 15:34:09
Time (end):   Wed, 2019-02-20 16:23:09
Duration: 49 minutes 0.40 seconds
Number of files: 1019883
Utilization of max. archive size: 0%
------------------------------------------------------------------------------
                       Original size      Compressed size    Deduplicated size
This archive:               58.48 GB             28.39 GB            133.29 MB
All archives:                9.94 TB              4.61 TB            111.96 GB

                       Unique chunks         Total chunks
Chunk index:                 1426888            153218526
------------------------------------------------------------------------------
```

**Comment by jumper444** [comment]
Presumable the sync_file_range line/command was(is) in the program for a reason.  I concur that it throws a WSL error since not implemented and that people are bypassing the error (now in experimental build) by commenting out the line.   This seems rather casual without an explanation as to why that line is there in the first place and what effects may occur by removing it.  Can anyone elaborate?

**Comment by ThomasWaldmann** [comment]
@jumper444 since borg 1.1.9 that code is not executed on WSL any more, no need to modify the borg code.

this does not affect data safety, the sync file range was just to get a less "bursty" behaviour (OS caching lots of write data and then writing it all at once to disk)

**Comment by laurensdijkstra** [comment]
The binary borg-linux64, version 1.1.9, that I downloaded from https://github.com/borgbackup/borg/releases still has the same issue. Am I missing some environment variables that make borg recognize it's WSL?

**Comment by ThomasWaldmann** [comment]
@laurensdijkstra https://github.com/borgbackup/borg/blob/1.1.9/src/borg/platform/linux.pyx#L243

This is the code for that.

Can you manually check whether you see `Microsoft` in `/proc/version` (use the same user as when running borg).

**Comment by theChaosCoder** [comment]
Output in "ubuntu 18.04" WSL
```
cat /proc/version
Linux version 4.4.0-17763-Microsoft (Microsoft@Microsoft.com) (gcc version 5.4.0 (GCC) ) #379-Microsoft Wed Mar 06 19:16:00 PST 2019
```

**Comment by laurensdijkstra** [comment]
I'm sorry, I copy pasted the command from earlier, when I was using the `borg` command. Replacing with ` borg-linux65` makes it work :)

**Comment by ThomasWaldmann** [comment]
@laurensdijkstra oh. so you are saying the 1.1.9 code for WSL works for you?

**Comment by theChaosCoder** [comment]
I just tested a bord init + create on local hdd, no errors in WSL.

**Comment by laurensdijkstra** [comment]
Yes

**Comment by NakedFerret** [comment]
Borg exploded when I tried to create an archive

```
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/borg/archiver.py", line 4253, in main
    exit_code = archiver.run(args)
  File "/usr/lib/python3/dist-packages/borg/archiver.py", line 4185, in run
    return set_ec(func(args))
  File "/usr/lib/python3/dist-packages/borg/archiver.py", line 152, in wrapper
    return method(self, args, repository=repository, **kwargs)
  File "/usr/lib/python3/dist-packages/borg/archiver.py", line 547, in do_create
    create_inner(archive, cache)
  File "/usr/lib/python3/dist-packages/borg/archiver.py", line 511, in create_inner
    read_special=args.read_special, dry_run=dry_run, st=st)
  File "/usr/lib/python3/dist-packages/borg/archiver.py", line 621, in _process
    read_special=read_special, dry_run=dry_run)
  File "/usr/lib/python3/dist-packages/borg/archiver.py", line 598, in _process
    status = archive.process_file(path, st, cache)
  File "/usr/lib/python3/dist-packages/borg/archive.py", line 1021, in process_file
    self.chunk_file(item, cache, self.stats, backup_io_iter(self.chunker.chunkify(fd, fh)))
  File "/usr/lib/python3/dist-packages/borg/archive.py", line 949, in chunk_file
    item.chunks.append(chunk_processor(data))
  File "/usr/lib/python3/dist-packages/borg/archive.py", line 937, in chunk_processor
    chunk_entry = cache.add_chunk(self.key.id_hash(data), data, stats, wait=False)
  File "/usr/lib/python3/dist-packages/borg/cache.py", line 898, in add_chunk
    self.repository.put(id, data, wait=wait)
  File "/usr/lib/python3/dist-packages/borg/repository.py", line 1072, in put
    segment, offset = self.io.write_put(id, data)
  File "/usr/lib/python3/dist-packages/borg/repository.py", line 1442, in write_put
    fd.write(b''.join((crc, header, id, data)))
  File "src/borg/platform/linux.pyx", line 255, in borg.platform.linux.SyncFile.write
  File "src/borg/platform/linux.pyx", line 229, in borg.platform.linux._sync_file_range
OSError: [Errno 38] Function not implemented

Platform: Linux DESKTOP-PM52I4Q 4.4.0-17763-Microsoft #864-Microsoft Thu Nov 07 15:22:00 PST 2019 x86_64
Linux: Ubuntu 18.04 bionic
Borg: 1.1.5  Python: CPython 3.6.9
PID: 26  CWD: /mnt/d
sys.argv: ['/usr/bin/borg', 'create', '--stats', '--progress', '--compression', 'lz4', 'RAW-backups-repo::b1', 'RAW-backups']
SSH_ORIGINAL_COMMAND: None
```

Seems related to this unimplemented system call
https://github.com/Microsoft/WSL/issues/645

**Comment by theChaosCoder** [comment]
INFO: The dir sync problem was fixed in WSL.

https://github.com/borgbackup/borg/issues/1961#issuecomment-335491180
https://github.com/microsoft/WSL/issues/2557

**Comment by arkanoid87** [comment]
It's 5+ year old issue, what's the general consensus over this and is this still to be consider experimental?

**Comment by ThomasWaldmann** [comment]
i personally do not use borg under WSL and there is no automated testing for it (like seen for other OSes in our Vagrantfile and on github actions CI), so from my PoV the state is unknown / experimental.

**Comment by bitinerant** [comment]
I've been running Borg under WSL for several years now. According to `borg list ...`, it runs reliably, though, I'm embarrassed to admit that I've never actually verified a backup. It's configured to launch each night via `schtasks /CREATE /TN \"Run Backup\" /RU Administrator /XML run-backup.xml /F` where the key section of the .xml file is:
```
  <Actions Context="Author">
    <Exec>
      <Command>C:\WSL\Ubuntu\ubuntu1804.exe</Command>
      <Arguments>run "/root/run-backup"</Arguments>
    </Exec>
  </Actions>
```
and `run-backup` launches Borg.

**Comment by arkanoid87** [comment]
> I've been running Borg under WSL for several years now. According to `borg list ...`, it runs reliably, though, I'm embarrassed to admit that I've never actually verified a backup. It's configured to launch each night via `schtasks /CREATE /TN \"Run Backup\" /RU Administrator /XML run-backup.xml /F` where the key section of the .xml file is:
> 
> ```
>   <Actions Context="Author">
>     <Exec>
>       <Command>C:\WSL\Ubuntu\ubuntu1804.exe</Command>
>       <Arguments>run "/root/run-backup"</Arguments>
>     </Exec>
>   </Actions>
> ```
> 
> and `run-backup` launches Borg.

Thank you for your feedback. I have a couple of questions:
- are you doing backup only for WSL files, or you get into `C:\` windows root fs too?
- why don't you try mounting a backup via fusefs and retrieving a file?

**Comment by mnpenner** [comment]
Fwiw I'm also running borg under WSL via windows task scheduler.

I did try restoring files a few times. I just used the cli borg commands,
no fuse business. Seems to work fine.

I back up both Windows and WSL files, and even some network files.

**Comment by bitinerant** [comment]
@arkanoid87 - I'm *only* backing up Windows files on those laptops. WSL is not used other than for Borg. As for not testing the backups, it just hasn't been a priority. And mounting with fusefs would take forever because the Borg server is a Raspberry Pi 3 in another time zone.

**Comment by jumper444** [comment]
I have looked up how to mount or FUSE a repository so I can access it like a storage system, but never found a process that seemed straight forward.  All my use has been command line.

**Comment by brummelte** [comment]
@jumper444 I backup using WSL 1 for improved speed and I use WSL 2 for restoring via FUSE mount. I can then even open the Linux FUSE folder in Windows Explorer.

**Comment by ThomasWaldmann** [comment]
As borg seems to work under WSL, shall we close that issue?

Stuff that should be documented should end up in the docs (PRs welcome) and if there is a specific issue, it can be brought up in a more specific ticket than this one.

**Comment by jumper444** [comment]
Thinking about it I have a comment/suggestion about the installation docs (for Borg on WSL) to put here for consideration.  It is late, please give me a day or two and not close yet.

**Comment by blue-cosmos** [comment]
I am trying Borg in WSL 1. All the data I am planning to backup are stored in NTFS volumes, Windows network shares and some virtual drives simulated by Dokany. I guess these filesystems cannot provide stable inodes when mounted in WSL. Is it safe to ignore inode in 'borg create' and rely on file path/ctime/size to detect file changes?

**Comment by ThomasWaldmann** [comment]
ctime is usually only controlled by the kernel (not by userspace), so it should be relatively safe.

be aware that there is a regression in borg 1.2.1 that makes repos fail on filesystems that do not support chmod (like smbfs/cifs, not sure about ntfs).

**Comment by ThomasWaldmann** [comment]
https://github.com/borgbackup/borg/issues/6786 see there.

**Comment by jumper444** [comment]
THOMAS: I was going to suggest (or write) a document change for "borg using WSL" installation section. Right now it says basically "go do what Ubuntu says" which is technically correct.  But we all know (I include myself) that many windows users aren't familiar with linux or package management or compiling or any of that.  We are given an EXE.  Download, install..and run.

My point and suggestion is that a small change in that documentation section would allow that instead - thus creating easy windows use (they can get more complex later if they want and as skills permit).  An easy installation/run section is possible because WSL is a rather 'fixed' and known distribution...managed and configured by Microsoft.  So it *IS* possible to just give a windows user an executable, tell them to 'put it here', and then tell them to just type "./borg" to run (no chmod, etc, is necessary).

That was going to be my suggestion for the "borg on wsl" section instead of "just go do ubuntu instructions".
Thoughts?

**Comment by jumper444** [comment]
My proposed change in the section would be:
1) go get this exact download via browser (it would be a link to the exact location of the 'borg-linux64' file; there would be no instructions about compiling, package managers, etc)
2) run WSL on your computer
3) inside WSL move (mv) the downloaded file to your home directory (with the exact command shown)
4) run borg by typing "./borg-linux64'

thus, a windows user on WSL is up and running in a minute or two just like with other windows applications instead of digging into 'go do ubuntu' which (to new people) is not as clear.

**Comment by jumper444** [comment]
That documentation section could still say "but you can also use the Ubuntu procedures as well" for more advanced users.

**Comment by jumper444** [comment]
I will write exact text for that proposed change if you agree and want to add it. 
If you want to just keep it as "go follow Ubuntu" then perfectly fine and the WSL topic and issue closure is fine/done with me.

Sorry for the multiple messages here.

**Comment by ThomasWaldmann** [comment]
You could make a PR against 1.2-maint branch. Master will soon be borg 2, so guess that's still too unstable for a docs change right now.

**Comment by ThomasWaldmann** [comment]
OK, guess I am closing this now. If anything notable comes up regarding borg on WSL, just open a new ticket or pull request.

**Comment by jbrepogmailcom** [comment]
Hello. I have separate linux partition. I boot into that and mount ntfs windows system with ntfs-3g. Then I run following command to backup windows except core installation, that I can recreate. I also exclude Steam games, because I save gamesaves to cloud and otherwise game installation can be recreated. The command has been recommended by ChatGPT and is running fine:

find /media/ntfs_partition -type f -newermt '2024-07-27' | borg create --compression zstd --stats --progress \
  --paths-from-stdin \
  --exclude '/media/ntfs_partition/$RECYCLE.BIN' \
  --exclude '/media/ntfs_partition/System Volume Information' \
  --exclude '/media/ntfs_partition/Windows/Temp' \
  --exclude '/media/ntfs_partition/Users/*/AppData/Local/Temp' \
  --exclude '/media/ntfs_partition/Users/*/AppData/Local/Microsoft/Windows/INetCache' \
  --exclude '/media/ntfs_partition/Users/*/AppData/Local/CrashDumps' \
  --exclude '/media/ntfs_partition/Windows/Prefetch' \
  --exclude '/media/ntfs_partition/Program Files (x86)/Steam' \
  --exclude-caches \
  <user>@<IP>:/RAID3T/BACKUPS/Borg/W10_partition::W10_backup_$(date +%Y-%m-%d)

**Comment by ThomasWaldmann** [comment]
@jbrepogmailcom thanks for the feedback.

I would not suggest using `find -newermt`, but just give borg the starting directory and let it do the fs recursion by itself and backup **ALL** the files in there minus the excluded ones.

The first backup will take a while, but the next one should be faster (if it is not, see our faq about the files cache and tune it accordingly).
